File: LeetCodeTasks/30_substring-with-concatenation-of-all-words/solutions/java/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (words.length == 0 || s.length() == 0) return result;

        int wordLength = words[0].length();
        int totalWordsLength = wordLength * words.length;
        
        if (totalWordsLength > s.length()) return result;
        
        Map<String, Integer> wordCount = new HashMap<>();
        for (String word : words) {
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }
        
        for (int i = 0; i < wordLength; ++i) {
            int left = i;
            Map<String, Integer> seenWords = new HashMap<>();
            int count = 0;
            
            for (int j = i; j <= s.length() - wordLength; j += wordLength) {
                String currentWord = s.substring(j, j + wordLength);
                
                if (wordCount.containsKey(currentWord)) {
                    seenWords.put(currentWord, seenWords.getOrDefault(currentWord, 0) + 1);
                    count++;
                    
                    while (seenWords.get(currentWord) > wordCount.get(currentWord)) {
                        String leftWord = s.substring(left, left + wordLength);
                        seenWords.put(leftWord, seenWords.get(leftWord) - 1);
                        count--;
                        left += wordLength;
                    }
                    
                    if (count == words.length) {
                        result.add(left);
                        String leftWord = s.substring(left, left + wordLength);
                        seenWords.put(leftWord, seenWords.get(leftWord) - 1);
                        count--;
                        left += wordLength;
                    }
                } else {
                    seenWords.clear();
                    count = 0;
                    left = j + wordLength;
                }
            }
        }
        return result;
    }
}

File: LeetCodeTasks/30_substring-with-concatenation-of-all-words/solutions/python3/solution.py
from typing import List
from collections import Counter

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not words or not s:
            return []

        word_length = len(words[0])
        total_words_length = len(words) * word_length
        word_count = Counter(words)
        result_indices = []

        for i in range(word_length):
            left = i
            current_count = Counter()

            for right in range(i, len(s) - word_length + 1, word_length):
                word = s[right:right + word_length]
                current_count[word] += 1

                while current_count[word] > word_count[word]:
                    current_count[s[left:left + word_length]] -= 1
                    left += word_length

                if (right + word_length - left) == total_words_length:
                    result_indices.append(left)

        return result_indices
